app/services/samples/transfer_service.rb

```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+module Samples
+  # Service used to Transfer Samples
+  class TransferService < BaseProjectService
+    TransferError = Class.new(StandardError)
+
+    def execute(sample_transfer)
+      @sample_transfer = sample_transfer
+
+      return false unless @sample_transfer
+
+      validate(@sample_transfer)
+
+      # Authorize if user can transfer samples from the current project
+      authorize! @project, to: :transfer_sample?
+
+      # Authorize if user can transfer samples to the new project
+      @new_project = Project.find_by(id: @sample_transfer.project_id)
+      authorize! @new_project, to: :transfer_sample_into_project?
+
+      transfer(@sample_transfer)
+
+      true
+    rescue Samples::TransferService::TransferError => e
+      sample_transfer.errors.add(:project_id, e.message)
+      false
+    end
+
+    private
+
+    def validate(sample_transfer)
+      @sample_transfer = sample_transfer
+      return unless @project.id == @sample_transfer.project_id.to_i
+
+      raise TransferError,
+            I18n.t('services.samples.transfer.duplicate_project')
+    end
+
+    def transfer(sample_transfer)
+      @sample_transfer = sample_transfer
+      ActiveRecord::Base.transaction do
+        JSON.parse(@sample_transfer.sample_ids.first).each do |sample_id|
```

Comment:
Is there a specific reason for using JSON.parse here? We should just be able to pass an array of ids from the controller and then iterate over it without having to use JSON.parse.

Also, can be add a check to make sure the destination project doesn't already contain a sample with the same name as the sample being transferred.

We probably also want to keep track of sample transfers that have errored and once the loop is done to raise a `TransferError`

Reply:
This has a little story behind it, lol.
In my first implementation I had wrapped the sample table in a form to send the selected samples to the next step. This didn't work, unless I commented out the action buttons that are available for each sample row. I don't know the exact cause, but stimulus broke the form. This is why the selection controller was implemented. I am populating a hidden field in the form that calls the transfer service. The form is sending that data in the following format:
![image](https://github.com/phac-nml/irida-next/assets/325703/cb91d1ed-3233-4954-97cb-0e2b0e345a5e)
I don't know how else to populate an array in a hidden field with JavaScript.

Reply:
Oh I see. I will take a look at that. Which file is it in?

Reply:
I populate the hidden field with the form_controller.js & app/views/projects/samples_transfer/new.html from within the transfer_samples branch.